src/granite/comparator.cr

```diff
@@ -0,0 +1,32 @@
+# Adds equals, hash, and comparison methods to Granite::Base objects.
+module Granite::Comparator
+  macro __process_comparator
+
+    def_equals_and_hash {{ *FIELDS.keys }}
+
+    {%
+      pk_type = PRIMARY[:type].id
+      pk_type_node = PRIMARY[:type].resolve
+      pk_field = PRIMARY[:name].id
+    %}
+
+    {% if pk_type_node.has_method?("<=>") %}
+    # Compares this `{{@type}}` to another `{{@type}}` by comparing their primary keys (`{{pk_field}}`).
+    #
+    # Returns 0 if the two primary keys are equal, a negative number if this object's primary key is considered less than other's, or a positive number otherwise.
+    {% else %}
+    # Raises an exception; the `{{pk_field}}` field does not implement the `<=>` operator.
+    {% end %}
+    def <=>(other : {{@type}})
+      raise "Cannot compare two {{@type}} objects if either's `{{pk_field}}` is `nil`." if self.{{pk_field}}.nil? || other.{{pk_field}}.nil?
+      {% if pk_type_node.has_method?("<=>") %}
+        self.{{pk_field}}.as({{pk_type}}) <=> other.{{pk_field}}.as({{pk_type}})
```

Comment:
Comparing primary keys doesn't seem like it would provide a good determination if something is greater than or less than another.  This would be something that would need the business context to know how to perform this.  I agree with @c910335 that we need to allow the users to override this.

I'm going to close this PR since I believe it needs some rethinking.  Thanks for your contribution.